fendermint/vm/interpreter/src/fvm/exec.rs

```diff
@@ -129,45 +129,49 @@ where
-        let updates = if let Some((checkpoint, power_table, updates)) =
+        let updates = if let Some((checkpoint, _, updates)) =
             checkpoint::maybe_create_checkpoint(&self.client, &self.gateway, &mut state)
                 .await
                 .context("failed to create checkpoint")?
         {
             // Asynchronously broadcast signature, if validating.
             if let Some(ref ctx) = self.validator_ctx {
-                if let Some(validator) = power_table
-                    .0
-                    .iter()
-                    .find(|v| v.public_key.0 == ctx.public_key)
-                    .cloned()
-                {
-                    // Do not resend past signatures.
-                    if !self.syncing().await? {
-                        let secret_key = ctx.secret_key.clone();
-                        let broadcaster = ctx.broadcaster.clone();
-                        let gateway = self.gateway.clone();
-                        let chain_id = state.chain_id();
-
-                        tokio::spawn(async move {
-                            let height = checkpoint.block_height;
-
-                            let res = checkpoint::broadcast_signature(
-                                &broadcaster,
-                                &gateway,
-                                checkpoint,
-                                &power_table,
-                                &validator,
-                                &secret_key,
-                                chain_id,
-                            )
-                            .await;
-
-                            if let Err(e) = res {
-                                tracing::error!(error =? e, height, "error broadcasting checkpoint signature");
-                            }
-                        });
-                    }
+                // Do not resend past signatures.
+                if !self.syncing().await? {
+                    // Fetch any incomplete checkpoints synchronously because the state can't be shared across threads.
+                    let incomplete_checkpoints = self
+                        .gateway
+                        .incomplete_checkpoints(&mut state)
+                        .context("failed to fetch incomplete checkpoints")?;
+
+                    debug_assert!(
+                        incomplete_checkpoints
+                            .iter()
+                            .any(|cp| cp.block_height == checkpoint.block_height
+                                && cp.block_hash == checkpoint.block_hash),
+                        "the current checkpoint is incomplete"
+                    );
+
+                    let client = self.client.clone();
+                    let gateway = self.gateway.clone();
+                    let chain_id = state.chain_id();
+                    let height = checkpoint.block_height;
+                    let validator_ctx = ctx.clone();
+
+                    tokio::spawn(async move {
```

Comment:
Why is it possibel to do the broadcast of incomplete signatures in parallel here without resulting in nonce races? (I also ask because in the comment above you mention that broadcasts can't be done in parallel).

Reply:
Sorry, it looks like my comment wasn't as clear as it could have been. 

What I meant was that we should not try to kick off a separate background task for each pending checkpoint like this:
```rust
for cp in pending_checkpoints {
  tokio::spawn(async move {
    checkpoint::broadcast_signature(broadcaster, cp).await;
  });
}
```
That's because every time the `broadcaster` is asked to send a transaction it will query the state for the nonce, and in this case that will surely mean all of the checkpoint submissions will get the same nonce and only one will go through. 

Instead, what we have is this:
```rust
for cp in pending_checkpoints {

    checkpoint::broadcast_signature(broadcaster, cp).await;
  });
}
```